packages/grafana-ui/src/components/uPlot/utils.ts

```diff
@@ -152,6 +149,196 @@ export function collectStackingGroups(f: Field, groups: Map<string, number[]>, s
   }
 }
 
+export interface StackingGroup {
+  series: number[];
+  dir: number;
+}
+
+// generates bands between adjacent group series
+export function getStackingBands(group: StackingGroup) {
+  let bands: uPlot.Band[] = [];
+  let { series, dir } = group;
+  let lastIdx = series.length - 1;
+
+  let rSeries = series.slice().reverse();
+
+  rSeries.forEach((si, i) => {
+    if (i !== lastIdx) {
+      let nextIdx = rSeries[i + 1];
+      bands.push({
+        series: [si, nextIdx],
+        // fill direction is inverted from stack direction
+        dir: -1 * dir,
+      });
+    }
+  });
+
+  return bands;
+}
+
+// expects an AlignedFrame
+export function getStackingGroups(frame: DataFrame) {
+  let groups: Map<string, StackingGroup> = new Map();
+
+  frame.fields.forEach(({ config, values }, i) => {
+    // skip x or time field
+    if (i === 0) {
+      return;
+    }
+
+    let { custom } = config;
+
+    if (custom == null) {
+      return;
+    }
+
+    // TODO: currently all AlignedFrame fields end up in uplot series & data, even custom.hideFrom?.viz
+    // ideally hideFrom.viz fields would be excluded so we can remove this
+    if (custom.hideFrom?.viz) {
+      return;
+    }
+
+    let { mode: stackingMode, group: stackingGroup } = custom.stacking;
+
+    // not stacking
+    if (stackingMode === StackingMode.None) {
+      return;
+    }
+
+    // will this be stacked up or down after any transforms applied
+    let vals = values.toArray();
+    let transform = custom.transform;
+    let stackDir =
+      transform === GraphTransform.Constant
+        ? vals[0] > 0
+          ? 1
+          : -1
+        : transform === GraphTransform.NegativeY
+        ? vals.some((v) => v > 0)
+          ? -1
+          : 1
+        : vals.some((v) => v > 0)
+        ? 1
+        : -1;
+
+    let drawStyle = custom.drawStyle;
+    let drawStyle2 =
+      drawStyle === GraphDrawStyle.Bars
+        ? custom.barAlignment
+        : drawStyle === GraphDrawStyle.Line
+        ? custom.lineInterpolation
+        : '';
+
+    let stackKey = [stackDir, stackingMode, stackingGroup, buildScaleKey(config), drawStyle, drawStyle2].join('|');
+
+    let group = groups.get(stackKey);
+
+    if (group == null) {
+      group = {
+        series: [],
+        dir: stackDir,
+      };
+
+      groups.set(stackKey, group);
+    }
+
+    group.series.push(i);
+  });
+
+  console.log(...groups.values());
+
+  return [...groups.values()];
+}
+
+/** @internal */
+export function preparePlotData2(
```

Comment:
Tests for this would also be good.

Reply:
ported most of the tests over from `preparePlotData()`. probably can write a few more for this variant since it does auto-grouping by pos/neg and by rendering style (drawStyle + barAlignment + lineInterpolation)

Reply:
i think the additional behavior is now quasi-covered by the `getStackingGroups()` tests